src/basic/audit-util.c

```diff
@@ -68,6 +70,67 @@ int audit_loginuid_from_pid(pid_t pid, uid_t *uid) {
         return 0;
 }
 
+static int try_audit_request(int fd) {
+        struct iovec iov;
+        struct msghdr mh;
+        ssize_t n;
+
+        assert(fd >= 0);
+
+        struct {
+                struct nlmsghdr hdr;
+                struct nlmsgerr err;
+        } _packed_ msg = {
+                .hdr.nlmsg_len = NLMSG_LENGTH(0),
+                .hdr.nlmsg_type = AUDIT_GET_FEATURE,
+                .hdr.nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK,
+        };
+        iov = (struct iovec) {
+                .iov_base = &msg,
+                .iov_len = msg.hdr.nlmsg_len,
+        };
+        mh = (struct msghdr) {
+                .msg_iov = &iov,
+                .msg_iovlen = 1,
+        };
+
+        if (sendmsg(fd, &mh, MSG_NOSIGNAL) < 0) {
+                log_debug_errno(errno, "Failed to send AUDIT_GET_FEATURE request, ignoring: %m");
+                return 0;
+        }
+
+        iov.iov_len = sizeof(msg);
+
+        n = recvmsg_safe(fd, &mh, 0);
+        if (n < 0) {
+                log_debug_errno(errno, "Failed to recv AUDIT_GET_FEATURE request ack, ignoring: %m");
+                return 0;
+        }
+
+        if (n != NLMSG_LENGTH(sizeof(struct nlmsgerr))) {
+                log_debug("Unexpected size for netlink ack, ignoring.");
+                return 0;
+        }
+
+        if (msg.hdr.nlmsg_type != NLMSG_ERROR) {
+                log_debug("Expected NLMSG_ERROR message but got %d, ignoring.", msg.hdr.nlmsg_type);
+                return 0;
+        }
+
+        if (msg.err.error == 0)
```

Comment:
so, usually we arrange our code in a way that the "clean"/"success" codepaths is not indented, and the "exceptional"/"error" codepath is indented. That would suggest you arrange things like this:

```c
if (msg.err.error == ECONNREFUSED) 
       return …
if (msg.err.error != 0
       return …

return 1;
```

i think this would be a bit nicer to follow